runtime/commonMain/src/kotlinx/serialization/SerialKinds.kt

```diff
@@ -0,0 +1,236 @@
+ * ```
+ * and its corresponding [Decoder] counterpart.
+ *
+ * ### Serial descriptor implementors note
+ * These kinds can be used not only for collection and regular classes.
+ * For example, provided serializer for [Map.Entry] represents it as [Map] type, so it is serialized
+ * as `{"actualKey": "actualValue"}` map directly instead of `{"key": "actualKey", "value": "actualValue"}`
+ */
+public sealed class StructureKind : SerialKind() {
+
+    /**
+     * Structure kind for regular classes with an arbitrary, but known statically, structure.
+     * Serializers typically encode classes with calls to [Encoder.beginStructure] and [CompositeEncoder.endStructure],
+     * writing the elements of the class between these calls.
+     */
+    public object CLASS : StructureKind()
+
+    /**
+     * Structure kind for lists of an arbitrary length.
+     * Serializers typically encode classes with calls to [Encoder.beginCollection] and [CompositeEncoder.endStructure],
+     * writing the elements of the list between these calls.
+     * Built-in list serializers treat elements as homogeneous, though application-specific serializers may impose
+     * application-specific restrictions on specific [LIST] types.
+     *
+     * Example of such application-specific serialization may be class `class ListOfThreeElements() : List<Any>`,
+     * for which an author of the serializer knows that while it is `List<Any>`, in fact, is always has three elements
+     * of a known type (e.g. the first is always a string, the second one is always an int etc.)
+     */
+    public object LIST : StructureKind()
+
+    /**
+     * Structure kind for maps of an arbitrary length.
+     * Serializers typically encode classes with calls to [Encoder.beginCollection] and [CompositeEncoder.endStructure],
+     * writing the elements of the map between these calls.
+     *
+     * Built-in map serializers treat elements as homogeneous, though application-specific serializers may impose
+     * application-specific restrictions on specific [MAP] types.
+     */
+    public object MAP : StructureKind()
+}
+
+/**
+ * Union structure kind represents a [tagged union][https://en.wikipedia.org/wiki/Tagged_union] structure,
+ * meaning that the type is represent by one of a multiple possible values (potentially unknown).
+ *
+ * An example of such union kind can be enum or its derivatives, such as "one of known strings".
+ */
+public sealed class UnionKind : SerialKind() {
+
+    // No documentation, let's decide whether it is a union at all
+    public object OBJECT : UnionKind()
+
+    /**
+     * Represents a Kotlin [Enum] with statically known values.
+     * All enum values should be enumerated in descriptor elements.
+     * Corresponding encoder and decoder methods are [Encoder.encodeEnum] and [Decoder.decodeEnum].
+     */
+    public object ENUM_KIND : UnionKind() // https://github.com/JetBrains/kotlin-native/issues/1447
```

Comment:
Also, the fact that `EnumDescriptor` has elements of ObjectKind should be mentioned somewhere